File: georocket-server/src/test/java/io/georocket/output/MergeNamespacesStrategyTest.java
package io.georocket.output;

import java.util.Arrays;

import org.junit.Test;
import org.junit.runner.RunWith;

import io.georocket.BufferWriteStream;
import io.georocket.SimpleChunkReadStream;
import io.georocket.storage.ChunkMeta;
import io.georocket.util.XMLStartElement;
import io.vertx.core.buffer.Buffer;
import io.vertx.ext.unit.Async;
import io.vertx.ext.unit.TestContext;
import io.vertx.ext.unit.junit.VertxUnitRunner;

/**
 * Test {@link MergeNamespacesStrategy}
 * @author Michel Kraemer
 */
@RunWith(VertxUnitRunner.class)
public class MergeNamespacesStrategyTest {
  private static final String XMLHEADER = "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n";
  
  private final static XMLStartElement ROOT1 = new XMLStartElement(null, "root",
      new String[] { "", "ns1", "xsi" },
      new String[] { "uri0", "uri1", "http://www.w3.org/2001/XMLSchema-instance" },
      new String[] { "xsi", "ns1" },
      new String[] { "schemaLocation", "attr1" },
      new String[] { "uri0 location0 uri1 location1", "value1" });
  private final static XMLStartElement ROOT2 = new XMLStartElement(null, "root",
      new String[] { "", "ns2", "xsi" },
      new String[] { "uri0", "uri2", "http://www.w3.org/2001/XMLSchema-instance" },
      new String[] { "xsi", "ns2" },
      new String[] { "schemaLocation", "attr2" },
      new String[] { "uri0 location0 uri2 location2", "value2" });
  private final static XMLStartElement EXPECTEDROOT = new XMLStartElement(null, "root",
      new String[] { "", "ns1", "xsi", "ns2" },
      new String[] { "uri0", "uri1", "http://www.w3.org/2001/XMLSchema-instance", "uri2" },
      new String[] { "xsi", "ns1", "ns2" },
      new String[] { "schemaLocation", "attr1", "attr2" },
      new String[] { "uri0 location0 uri1 location1 uri2 location2", "value1", "value2" });
  
  private final static String CONTENTS1 = "<elem><ns1:child1></ns1:child1></elem>";
  private final static Buffer CHUNK1 = Buffer.buffer(XMLHEADER + ROOT1 + CONTENTS1 + "</" + ROOT1.getName() + ">");
  private final static String CONTENTS2 = "<elem><ns2:child2></ns2:child2></elem>";
  private final static Buffer CHUNK2 = Buffer.buffer(XMLHEADER + ROOT2 + CONTENTS2 + "</" + ROOT2.getName() + ">");
  
  private final static ChunkMeta META1 = new ChunkMeta(Arrays.asList(ROOT1),
      XMLHEADER.length() + ROOT1.toString().length(),
      CHUNK1.length() - ROOT1.getName().length() - 3);
  private final static ChunkMeta META2 = new ChunkMeta(Arrays.asList(ROOT2),
      XMLHEADER.length() + ROOT2.toString().length(),
      CHUNK2.length() - ROOT2.getName().length() - 3);
  
  /**
   * Test a simple merge
   * @param context the test context
   */
  @Test
  public void simple(TestContext context) {
    Async async = context.async();
    MergeStrategy strategy = new MergeNamespacesStrategy();
    BufferWriteStream bws = new BufferWriteStream();
    strategy.init(META1, context.asyncAssertSuccess(v1 -> {
      strategy.init(META2, context.asyncAssertSuccess(v2 -> {
        strategy.merge(new SimpleChunkReadStream(CHUNK1), META1, bws, context.asyncAssertSuccess(v3 -> {
          strategy.merge(new SimpleChunkReadStream(CHUNK2), META2, bws, context.asyncAssertSuccess(v4 -> {
            strategy.finishMerge(bws);
            context.assertEquals(XMLHEADER + EXPECTEDROOT + CONTENTS1 + CONTENTS2 + "</" + EXPECTEDROOT.getName() + ">",
                bws.getBuffer().toString("utf-8"));
            async.complete();
          }));
        }));
      }));
    }));
  }
  
  /**
   * Make sure that chunks that have not been passed to the initalize method cannot be merged
   * @param context the test context
   */
  @Test
  public void mergeUninitialized(TestContext context) {
    Async async = context.async();
    MergeStrategy strategy = new MergeNamespacesStrategy();
    BufferWriteStream bws = new BufferWriteStream();
    strategy.init(META1, context.asyncAssertSuccess(v1 -> {
      // skip second init
      strategy.merge(new SimpleChunkReadStream(CHUNK1), META1, bws, context.asyncAssertSuccess(v3 -> {
        strategy.merge(new SimpleChunkReadStream(CHUNK2), META2, bws, context.asyncAssertFailure(err -> {
          async.complete();
        }));
      }));
    }));
  }
}

File: src/test/kotlin/io/georocket/output/xml/MergeNamespacesStrategyTest.kt
package io.georocket.output.xml

import io.georocket.assertThatThrownBy
import io.georocket.coVerify
import io.georocket.storage.GenericXmlChunkMeta
import io.georocket.util.XMLStartElement
import io.georocket.util.io.BufferWriteStream
import io.vertx.core.Vertx
import io.vertx.core.buffer.Buffer
import io.vertx.junit5.VertxExtension
import io.vertx.junit5.VertxTestContext
import io.vertx.kotlin.coroutines.dispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

/**
 * Test [MergeNamespacesStrategy]
 * @author Michel Kraemer
 */
@ExtendWith(VertxExtension::class)
class MergeNamespacesStrategyTest {
  companion object {
    private const val XMLHEADER = """<?xml version="1.0" encoding="UTF-8" standalone="yes"?>${"\n"}"""

    private val ROOT1 = XMLStartElement(null, "root",
      namespacePrefixes = listOf(null, "ns1", "xsi"),
      namespaceUris = listOf("uri0", "uri1", "http://www.w3.org/2001/XMLSchema-instance"),
      attributePrefixes = listOf("xsi", "ns1"),
      attributeLocalNames = listOf("schemaLocation", "attr1"),
      attributeValues = listOf("uri0 location0 uri1 location1", "value1")
    )
    private val ROOT2 = XMLStartElement(
      null, "root",
      namespacePrefixes = listOf(null, "ns2", "xsi"),
      namespaceUris = listOf("uri0", "uri2", "http://www.w3.org/2001/XMLSchema-instance"),
      attributePrefixes = listOf("xsi", "ns2"),
      attributeLocalNames = listOf("schemaLocation", "attr2"),
      attributeValues = listOf("uri0 location0 uri2 location2", "value2")
    )
    private val EXPECTEDROOT = XMLStartElement(
      null, "root",
      namespacePrefixes = listOf(null, "ns1", "xsi", "ns2"),
      namespaceUris = listOf("uri0", "uri1", "http://www.w3.org/2001/XMLSchema-instance", "uri2"),
      attributePrefixes = listOf("xsi", "ns1", "ns2"),
      attributeLocalNames = listOf("schemaLocation", "attr1", "attr2"),
      attributeValues = listOf("uri0 location0 uri1 location1 uri2 location2", "value1", "value2")
    )

    private const val CONTENTS1 = "<elem><ns1:child1></ns1:child1></elem>"
    private val CHUNK1 = Buffer.buffer(CONTENTS1)
    private const val CONTENTS2 = "<elem><ns2:child2></ns2:child2></elem>"
    private val CHUNK2 = Buffer.buffer(CONTENTS2)

    private val META1 = GenericXmlChunkMeta(listOf(ROOT1))
    private val META2 = GenericXmlChunkMeta(listOf(ROOT2))
  }

  /**
   * Test a simple merge
   */
  @Test
  fun simple(vertx: Vertx, ctx: VertxTestContext) {
    val strategy = MergeNamespacesStrategy()
    val bws = BufferWriteStream()

    CoroutineScope(vertx.dispatcher()).launch {
      ctx.coVerify {
        strategy.init(META1)
        strategy.init(META2)
        strategy.merge(CHUNK1, META1, bws)
        strategy.merge(CHUNK2, META2, bws)
        strategy.finish(bws)
        assertThat(bws.buffer.toString("utf-8")).isEqualTo(
            """$XMLHEADER$EXPECTEDROOT$CONTENTS1$CONTENTS2</${EXPECTEDROOT.name}>""")
      }
      ctx.completeNow()
    }
  }

  /**
   * Make sure that chunks that have not been passed to the initalize method cannot be merged
   */
  @Test
  fun mergeUninitialized(vertx: Vertx, ctx: VertxTestContext) {
    val strategy = MergeNamespacesStrategy()
    val bws = BufferWriteStream()

    CoroutineScope(vertx.dispatcher()).launch {
      ctx.coVerify {
        assertThatThrownBy {
          strategy.init(META1) // skip second init
          strategy.merge(CHUNK1, META1, bws)
          strategy.merge(CHUNK2, META2, bws)
        }.isInstanceOf(IllegalStateException::class.java)
      }
      ctx.completeNow()
    }
  }
}
